src/index.js

```diff
@@ -1,16 +1,11 @@
 function createActionEnhancerMiddleware(mappings) {
   return ({getState}) => next => action => {
     const enhancedFields = mappings.reduce((current, nextMapping) => {
-      let result = current;
-      if (action[nextMapping.id]) {
-        result = Object.assign({}, result);
-
-        if (nextMapping.mapState) {
-          Object.assign(result, nextMapping.mapState(getState()));
-        }
+      if (nextMapping.id in action) {
+        Object.assign(current, nextMapping.mapState(getState(), action[nextMapping.id]));
```

Comment:
the key is that i now pass action[nextMapping.id] to mapState as param 2. it's backwards compatible